### ISSUE TITLE
List.hs doctest with base-4.8.0.0

### ISSUE THREAD

**Issue by zaquest** [NICTA/course#131]
Since `base-4.8.0.0` type of `Prelude.foldr` changed from `(a -> b -> b) -> b -> [a] -> b` to `Foldable t => (a -> b -> b) -> b -> t a -> b` that causes an issue with doctests in `src/Course/List.hs` file
```
> doctest -isrc -Wall -fno-warn-type-defaults src/Course/List.hs

src/Course/Core.hs:50:1: Warning:
    The import of ‘Char’ from module ‘Prelude’ is redundant
src/Course/Core.hs:50:1: Warning:
    The import of ‘Char’ from module ‘Prelude’ is redundant
### Failure in src/Course/List.hs:28: expression `instance Arbitrary a => Arbitrary (List a) where arbitrary = P.fmap (P.foldr (:.) Nil) arbitrary'
expected:
 but got:
          <interactive>:26:70:
              Could not deduce (P.Foldable t0) arising from a use of ‘P.foldr’
              from the context (Arbitrary a)
                bound by the instance declaration at <interactive>:26:10-42
              The type variable ‘t0’ is ambiguous
              Note: there are several potential instances:
                instance P.Foldable (A.Const m) -- Defined in ‘Control.Applicative’
                instance P.Foldable (Either a) -- Defined in ‘Data.Foldable’
                instance P.Foldable Data.Proxy.Proxy -- Defined in ‘Data.Foldable’
                ...plus five others
              In the first argument of ‘P.fmap’, namely ‘(P.foldr (:.) Nil)’
              In the expression: P.fmap (P.foldr (:.) Nil) arbitrary
              In an equation for ‘arbitrary’:
                  arbitrary = P.fmap (P.foldr (:.) Nil) arbitrary

          <interactive>:26:88:
              Could not deduce (Arbitrary (t0 a))
                arising from a use of ‘arbitrary’
              from the context (Arbitrary a)
                bound by the instance declaration at <interactive>:26:10-42
              The type variable ‘t0’ is ambiguous
              Relevant bindings include
                arbitrary :: Gen (List a) (bound at <interactive>:26:50)
              Note: there are several potential instances:
                instance [overlap ok] [safe] (P.RealFloat a, Arbitrary a) =>
                                             Arbitrary (Data.Complex.Complex a)
                  -- Defined in ‘Test.QuickCheck.Arbitrary’
                instance [overlap ok] [safe] (Arbitrary a, Arbitrary b) =>
                                             Arbitrary (Either a b)
                  -- Defined in ‘Test.QuickCheck.Arbitrary’
                instance [overlap ok] [safe] Data.Fixed.HasResolution a =>
                                             Arbitrary (Data.Fixed.Fixed a)
                  -- Defined in ‘Test.QuickCheck.Arbitrary’
                ...plus 9 others
              In the second argument of ‘P.fmap’, namely ‘arbitrary’
              In the expression: P.fmap (P.foldr (:.) Nil) arbitrary
              In an equation for ‘arbitrary’:
                  arbitrary = P.fmap (P.foldr (:.) Nil) arbitrary
Examples: 55  Tried: 4  Errors: 0  Failures: 1
```
The simple fix is to replace line
``` -- >>> instance Arbitrary a => Arbitrary (List a) where arbitrary = P.fmap (P.foldr (:.) Nil) arbitrary```
with
``` -- >>> instance Arbitrary a => Arbitrary (List a) where arbitrary = P.fmap ((P.foldr :: (a -> List a -> List a) -> List a -> [a] -> List a) (:.) Nil) arbitrary```
However, I am not sure if it's the best solution.

**Comment by zaquest** [comment]
Same problem in `ListZipper.hs` on line 24
```
-- >>> instance Arbitrary a => Arbitrary (List a) where arbitrary = P.fmap (P.foldr (:.) Nil) arbitrary
```

**Comment by zaquest** [comment]
Fixed in 8476d5a397716b13d3b61b9c53a046bc5ec5b4f8
